sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala

```diff
@@ -83,32 +83,37 @@ case class ApproximatePercentile(
   }
 
   // Mark as lazy so that accuracyExpression is not evaluated during tree transformation.
-  private lazy val accuracy: Int = accuracyExpression.eval().asInstanceOf[Int]
-
-  override def inputTypes: Seq[AbstractDataType] = {
-    // Support NumericType, DateType and TimestampType since their internal types are all numeric,
-    // and can be easily cast to double for processing.
-    Seq(TypeCollection(NumericType, DateType, TimestampType),
-      TypeCollection(DoubleType, ArrayType(DoubleType)), IntegerType)
-  }
+  private lazy val accuracy: Long = accuracyExpression.eval().asInstanceOf[Number].longValue()
 
   // Mark as lazy so that percentageExpression is not evaluated during tree transformation.
   private lazy val (returnPercentileArray: Boolean, percentages: Array[Double]) =
-    percentageExpression.eval() match {
-      // Rule ImplicitTypeCasts can cast other numeric types to double
-      case num: Double => (false, Array(num))
-      case arrayData: ArrayData => (true, arrayData.toDoubleArray())
+    percentageExpression.dataType match {
+      case DoubleType => (false, Array(percentageExpression.eval().asInstanceOf[Double]))
+      case _: NumericType =>
```

Comment:
Where does the overflow occur, before?

Reply:
here https://github.com/apache/spark/pull/26905/files#diff-7448088085a868386c669c69629c8169L86

Reply:
Why would it overflow? The type coercion should cast `accuracyExpression` to int type.

Reply:
If we specify a accuracy value greater than Int.MaxValue

Reply:
I don't mean a overflow exception here, just the value here is not right.

Reply:
shall we simply update `inputTypes` to replace `IntegerType` with `LongType`?

Reply:
The type coercion rule does not quite fit this function, in hive this func is implemented without implicit casting, each argument is strictly to its type and range. Our imlementation is not as same as our doc says

Reply:
We can do custom type coercion (don't extend ImplicitCastInputTypes) and disallow casting long to int implicitly for this function.

Reply:
Ok, I' ll check on this.

Reply:
BTW, as an aside, allowing very large values > 2 billion doesn't matter here. This is the inverse of an accuracy param and anything that large makes the resulting param basically 0 anyway, so, just do whatever is simpler IMHO, including disallowing long if that helps.